cpp/src/io/json/json_column.cu

```diff
@@ -0,0 +1,810 @@
+                  << (schema.has_value() ? std::to_string(static_cast<int>(schema->type.id()))
+                                         : "n/a");
+#endif
+        target_type = schema.value().type;
+      }
+      // Infer column type, if we don't have an explicit type for it
+      else {
+        target_type = cudf::io::detail::infer_data_type(
+          parsing_options(options).json_view(), d_input, string_ranges_it, col_size, stream);
+      }
+      // Convert strings to the inferred data type
+      auto col = experimental::detail::parse_data(string_spans_it,
+                                                  col_size,
+                                                  target_type,
+                                                  make_validity(json_col).first,
+                                                  parsing_options(options).view(),
+                                                  stream,
+                                                  mr);
+
+      // Reset nullable if we do not have nulls
+      // This is to match the existing JSON reader's behaviour:
+      // - Non-string columns will always be returned as nullable
+      // - String columns will be returned as nullable, iff there's at least one null entry
+      if (target_type.id() == type_id::STRING and col->null_count() == 0) {
+        col->set_null_mask(rmm::device_buffer{0, stream, mr}, 0);
+      }
+
+      // For string columns return ["offsets", "char"] schema
+      if (target_type.id() == type_id::STRING) {
+        return {std::move(col), {{"offsets"}, {"chars"}}};
+      }
+      // Non-string leaf-columns (e.g., numeric) do not have child columns in the schema
+      else {
+        return {std::move(col), {}};
+      }
+      break;
+    }
+    case json_col_t::StructColumn: {
+      std::vector<std::unique_ptr<column>> child_columns;
+      std::vector<column_name_info> column_names{};
+      size_type num_rows{json_col.num_rows};
+      // Create children columns
+      for (auto const& col_name : json_col.column_order) {
+        auto const& col = json_col.child_columns.find(col_name);
+        column_names.emplace_back(col->first);
+        auto& child_col            = col->second;
+        auto [child_column, names] = json_column_to_cudf_column2(
+          child_col, d_input, options, get_child_schema(col_name), stream, mr);
+        CUDF_EXPECTS(num_rows == child_column->size(),
+                     "All children columns must have the same size");
+        child_columns.push_back(std::move(child_column));
+        column_names.back().children = names;
+      }
+      auto [result_bitmask, null_count] = make_validity(json_col);
+      return {
+        make_structs_column(
+          num_rows, std::move(child_columns), null_count, std::move(result_bitmask), stream, mr),
+        column_names};
+      break;
+    }
+    case json_col_t::ListColumn: {
+      size_type num_rows = json_col.child_offsets.size();
```

Comment:
It would be more semantically consistent with cudf to use:

`size_type num_rows = json_col.child_offsets.size() - 1;`

and then add +1 as necessary later on.

Reply:
That's right. Updated.